rules/scripts/bitstreams_workspace.py

```diff
@@ -208,40 +211,50 @@ def _GetFromRemote(self, key):
         if latest:
             self.Touch(latest)
 
-    def GetFromCache(self, key):
+    def GetFromCache(self, key: str) -> Dict[str, Set[str]]:
         """Get the requested bitstream files.
 
         Args:
-            key: str; A git hash or the string 'latest'.
+            key: A git hash or the string 'latest'.
         Returns:
-            dict[str:str]: A dictionary mapping the bitstream types to
-                           their target files.
+            A dictionary that maps file extensions to the set of bitstream files
+            with that extension.
         """
         files = self._GetFromLocal(key)
         if not files:
             self._GetFromRemote(key)
             files = self._GetFromLocal(key)
-        return {os.path.splitext(f)[1][1:]: f for f in files}
 
-    def WriteBuildFile(self, build, key):
-        """Write a BUILD file for the requested bitstream files.
+        extension_map: Dict[str, Set[str]] = collections.defaultdict(set)
+        for file_name in files:
+            _, extension = os.path.splitext(file_name)
+            if extension[0] != '.':
+                logging.error("Bitstream file has no extension: " +
+                              repr(file_name))
+                sys.exit(1)
+            key = extension[1:]
+            extension_map[key] |= {file_name}
+        return extension_map
 
-        Args:
-            build: path; Filename of the BUILD file to write.
-            key: str; A git hash or the string 'latest'.
-        """
-        param = self.GetFromCache(key)
-        param['datetime'] = datetime.datetime.now().isoformat()
-        param['key'] = key
+    @staticmethod
+    def _GetDateTimeStr():
+        return datetime.datetime.now().isoformat()
 
-        if 'orig' not in param or 'splice' not in param:
+    def _ConstructBazelString(self, build_file: Path, key: str) -> str:
```

Comment:
Splitting this out means I can simplify the tests at the expense of code coverage for the actual file-writing function. Still not sure how I feel about mocking the `open()` function.

Reply:
I wouldn't worry about it.  The file writing function is now simple enough that it can be judged to be correct by inspection.

Reply:
Sounds good. Removing the `open()` mock. I also noticed that I was iterating over MMI files nondeterministically, which made the test flaky.